File: backend/routers/admin_reservation.py
from fastapi import APIRouter, Depends, HTTPException,WebSocket, WebSocketDisconnect
from sqlalchemy.orm import Session
from app.websocket_manager import manager
import json

from app.database import get_db
from app import models, schemas
from app.oauth2 import get_current_user


router = APIRouter(
    prefix="/admin_reservations",
    tags=["Admin Reservations"]
)


# ---------------------------------------------
@router.get("/")
def get_all_reservations(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):

    # =========================
    # ADMIN CHECK
    # =========================

    role = str(
        current_user.role
    ).strip().lower()

    if role != "admin":
        raise HTTPException(
            status_code=403,
            detail="Only admin can access reservations"
        )

    # =========================
    # GET RESERVATIONS
    # =========================

    reservations = (
        db.query(models.Reservation)
        .order_by(
            models.Reservation.id.desc()
        )
        .all()
    )

    result = []

    # =========================
    # LOOP RESERVATIONS
    # =========================

    for reservation in reservations:

        # -------------------------
        # USER
        # -------------------------

        user = (
            db.query(models.User)
            .filter(
                models.User.id ==
                reservation.user_id
            )
            .first()
        )

        # -------------------------
        # PARKING
        # -------------------------

        parking = (
            db.query(models.Parking)
            .filter(
                models.Parking.id ==
                reservation.parking_id
            )
            .first()
        )

        # -------------------------
        # SLOT
        # -------------------------

        slot = (
            db.query(models.ParkingSlot)
            .filter(
                models.ParkingSlot.id ==
                reservation.slot_id
            )
            .first()
        )

        # -------------------------
        # VEHICLE
        # -------------------------

        vehicle = None

        vehicle_id = getattr(
            reservation,
            "vehicle_id",
            None
        )

        if vehicle_id is not None:

            vehicle = (
                db.query(models.Vehicle)
                .filter(
                    models.Vehicle.id ==
                    vehicle_id
                )
                .first()
            )

        # =========================
        # SAFE USER DATA
        # =========================

        user_name = "-"

        if user:

            user_name = (
                getattr(user, "name", None)
                or getattr(user, "username", None)
                or getattr(user, "email", "-")
            )

        user_email = "-"

        if user:

            user_email = getattr(
                user,
                "email",
                "-"
            )

        # =========================
        # SAFE PARKING DATA
        # =========================

        parking_name = "-"

        parking_address = "-"

        if parking:

            parking_name = getattr(
                parking,
                "parking_name",
                "-"
            )

            parking_address = getattr(
                parking,
                "address",
                "-"
            )

        # =========================
        # SAFE SLOT DATA
        # =========================

        slot_number = "-"

        if slot:

            slot_number = (
                getattr(
                    slot,
                    "slot_number",
                    None
                )
                or getattr(
                    slot,
                    "slot_name",
                    None
                )
                or getattr(
                    slot,
                    "name",
                    None
                )
                or "-"
            )

        # =========================
        # SAFE VEHICLE DATA
        # =========================

        vehicle_number = "-"

        vehicle_type = "-"

        if vehicle:

            vehicle_number = getattr(
                vehicle,
                "vehicle_number",
                "-"
            )

            vehicle_type = getattr(
                vehicle,
                "vehicle_type",
                "-"
            )

        else:

            vehicle_number = getattr(
                reservation,
                "vehicle_number",
                "-"
            )

            vehicle_type = getattr(
                reservation,
                "vehicle_type",
                "-"
            )

        # =========================
        # RESULT
        # =========================

        result.append({

            "id": reservation.id,

            "user_name": user_name,

            "user_email": user_email,

            "parking_name": parking_name,

            "parking_address": parking_address,

            "slot_number": slot_number,

            "vehicle_number": vehicle_number,

            "vehicle_type": vehicle_type,

            "status": getattr(
                reservation,
                "status",
                None
            ),

            "reservation_date": getattr(
                reservation,
                "reservation_date",
                None
            ),

            "entry_time": getattr(
                reservation,
                "entry_time",
                None
            ),

            "exit_time": getattr(
                reservation,
                "exit_time",
                None
            ),

            "total_amount": getattr(
                reservation,
                "total_amount",
                None
            ),

            "created_at": getattr(
                reservation,
                "created_at",
                None
            )
        })

    return result

# -------------------view one reservation----------------

@router.get("/{reservation_id}")
def get_reservation(
    reservation_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):

    try:

        # =========================
        # ADMIN CHECK
        # =========================

        role = str(
            current_user.role
        ).strip().lower()

        if role != "admin":
            raise HTTPException(
                status_code=403,
                detail="Only admin can view reservations"
            )

        # =========================
        # GET RESERVATION
        # =========================

        reservation = (
            db.query(models.Reservation)
            .filter(
                models.Reservation.id ==
                reservation_id
            )
            .first()
        )

        if not reservation:

            raise HTTPException(
                status_code=404,
                detail="Reservation not found"
            )

        # =========================
        # USER
        # =========================

        user = (
            db.query(models.User)
            .filter(
                models.User.id ==
                reservation.user_id
            )
            .first()
        )

        # =========================
        # PARKING
        # =========================

        parking = (
            db.query(models.Parking)
            .filter(
                models.Parking.id ==
                reservation.parking_id
            )
            .first()
        )

        # =========================
        # SLOT
        # =========================

        slot = (
            db.query(models.ParkingSlot)
            .filter(
                models.ParkingSlot.id ==
                reservation.slot_id
            )
            .first()
        )

        # =========================
        # VEHICLE
        # =========================

        vehicle = None

        vehicle_id = getattr(
            reservation,
            "vehicle_id",
            None
        )

        if vehicle_id:

            vehicle = (
                db.query(models.Vehicle)
                .filter(
                    models.Vehicle.id ==
                    vehicle_id
                )
                .first()
            )

        # =========================
        # USER INFORMATION
        # =========================

        user_name = "-"

        user_email = "-"

        if user:

            user_name = (
                getattr(
                    user,
                    "name",
                    None
                )
                or getattr(
                    user,
                    "username",
                    None
                )
                or getattr(
                    user,
                    "email",
                    None
                )
                or "-"
            )

            user_email = getattr(
                user,
                "email",
                "-"
            )

        # =========================
        # PARKING INFORMATION
        # =========================

        parking_name = "-"

        parking_address = "-"

        parking_area = "-"

        parking_city = "-"

        if parking:

            parking_name = getattr(
                parking,
                "parking_name",
                "-"
            )

            parking_address = getattr(
                parking,
                "address",
                "-"
            )

            parking_area = getattr(
                parking,
                "area",
                "-"
            )

            parking_city = getattr(
                parking,
                "city",
                "-"
            )

        # =========================
        # SLOT INFORMATION
        # =========================

        slot_number = "-"

        slot_status = "-"

        if slot:

            slot_number = (
                getattr(
                    slot,
                    "slot_number",
                    None
                )
                or getattr(
                    slot,
                    "slot_name",
                    None
                )
                or getattr(
                    slot,
                    "name",
                    None
                )
                or "-"
            )

            slot_status = getattr(
                slot,
                "status",
                "-"
            )

        # =========================
        # VEHICLE INFORMATION
        # =========================

        vehicle_number = "-"

        vehicle_type = "-"

        if vehicle:

            vehicle_number = getattr(
                vehicle,
                "vehicle_number",
                "-"
            )

            vehicle_type = getattr(
                vehicle,
                "vehicle_type",
                "-"
            )

        else:

            vehicle_number = getattr(
                reservation,
                "vehicle_number",
                "-"
            )

            vehicle_type = getattr(
                reservation,
                "vehicle_type",
                "-"
            )

        # =========================
        # RESPONSE
        # =========================

        return {

            "id": reservation.id,

            "user": {
                "id": (
                    user.id
                    if user
                    else reservation.user_id
                ),
                "name": user_name,
                "email": user_email
            },

            "parking": {
                "id": (
                    parking.id
                    if parking
                    else reservation.parking_id
                ),
                "name": parking_name,
                "address": parking_address,
                "area": parking_area,
                "city": parking_city
            },

            "slot": {
                "id": (
                    slot.id
                    if slot
                    else reservation.slot_id
                ),
                "number": slot_number,
                "status": slot_status
            },

            "vehicle": {
                "number": vehicle_number,
                "type": vehicle_type
            },

            "status": getattr(
                reservation,
                "status",
                None
            ),

            "reservation_date": getattr(
                reservation,
                "reservation_date",
                None
            ),

            "entry_time": getattr(
                reservation,
                "entry_time",
                None
            ),

            "exit_time": getattr(
                reservation,
                "exit_time",
                None
            ),

            "total_amount": getattr(
                reservation,
                "total_amount",
                None
            ),

            "created_at": getattr(
                reservation,
                "created_at",
                None
            )
        }

    except HTTPException:
        raise

    except Exception as e:

        import traceback

        print(
            "VIEW RESERVATION ERROR:"
        )

        traceback.print_exc()

        raise HTTPException(
            status_code=500,
            detail=str(e)
        )


# ---------------------------------------
@router.websocket("/ws")
async def admin_reservation_websocket(
    websocket: WebSocket
):
    await manager.connect(websocket)

    print(
        "ADMIN RESERVATION WEBSOCKET CONNECTED"
    )

    try:
        while True:
            await websocket.receive_text()

    except WebSocketDisconnect:
        print(
            "ADMIN RESERVATION WEBSOCKET DISCONNECTED"
        )

        manager.disconnect(websocket)

# -----------------delete reservation-----------------
@router.delete("/{reservation_id}")
async def delete_reservation(
    reservation_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user)
):

    # ============================
    # CHECK ADMIN
    # ============================

    if str(current_user.role).strip().lower() != "admin":
        raise HTTPException(
            status_code=403,
            detail="Only admin can delete reservations"
        )

    # ============================
    # FIND RESERVATION
    # ============================

    reservation = (
        db.query(models.Reservation)
        .filter(
            models.Reservation.id == reservation_id
        )
        .first()
    )

    if not reservation:
        raise HTTPException(
            status_code=404,
            detail="Reservation not found"
        )

    try:

        # ============================
        # DELETE PAYMENT FIRST
        # ============================

        payments = (
            db.query(models.Payment)
            .filter(
                models.Payment.reservation_id == reservation_id
            )
            .all()
        )

        for payment in payments:
            db.delete(payment)

        # ============================
        # DELETE RESERVATION
        # ============================

        db.delete(reservation)

        # ============================
        # COMMIT
        # ============================

        db.commit()

        # ============================
        # WEBSOCKET
        # ============================

        await manager.broadcast(
            json.dumps({
                "event": "reservation_deleted",
                "data": {
                    "id": reservation_id
                }
            })
        )

        return {
            "message": "Reservation deleted successfully",
            "reservation_id": reservation_id
        }

    except Exception as e:

        db.rollback()

        print(
            "Delete reservation error:",
            str(e)
        )

        raise HTTPException(
            status_code=500,
            detail="Failed to delete reservation"
        )

File: frontend/js/admin_reservation.js
const API_BASE_URL =
    "https://smart-parking-system-tz4z.onrender.com";


let allReservations = [];

let socket = null;


// =====================================================
// TOKEN
// =====================================================

function getToken() {

    return sessionStorage.getItem(
        "access_token"
    );

}



// =====================================================
// LOAD RESERVATIONS
// =====================================================

async function loadReservations() {

    const token =
        getToken();


    if (!token) {

        console.error(
            "Admin token not found"
        );

        return;

    }


    try {

        const response =
            await fetch(
                `${API_BASE_URL}/admin_reservations/`,
                {
                    method: "GET",

                    headers: {
                        "Authorization":
                            `Bearer ${token}`,

                        "Content-Type":
                            "application/json"
                    }
                }
            );


        if (!response.ok) {

            console.error(
                "Reservation API error:",
                response.status
            );

            console.error(
                await response.text()
            );

            return;

        }


        const data =
            await response.json();


        console.log(
            "Admin reservations:",
            data
        );


        allReservations = data;


        displayReservations(
            allReservations
        );


    } catch (error) {

        console.error(
            "Error loading reservations:",
            error
        );

    }

}



// =====================================================
// DISPLAY RESERVATIONS
// =====================================================

function displayReservations(
    reservations
) {

    const tbody =
        document.getElementById(
            "reservationTableBody"
        );


    tbody.innerHTML = "";


    if (
        !reservations ||
        reservations.length === 0
    ) {

        tbody.innerHTML = `

            <tr>

                <td
                    colspan="7"
                    class="text-center text-muted"
                >

                    No reservations found

                </td>

            </tr>

        `;

        return;

    }


    reservations.forEach(
        reservation => {

            addReservationRow(
                reservation,
                tbody
            );

        }
    );

}



// =====================================================
// ADD RESERVATION ROW
// =====================================================

function addReservationRow(
    reservation,
    tbody
) {

    const row =
        document.createElement("tr");


    row.id =
        `reservation-row-${reservation.id}`;


    row.innerHTML = `

        <td>
            ${reservation.id ?? "-"}
        </td>

        <td>
            ${reservation.user_name ?? "-"}
        </td>

        <td>
            ${reservation.parking_name ?? "-"}
        </td>

        <td>
            ${reservation.slot_number ?? "-"}
        </td>

        <td>
            ${reservation.vehicle_number ?? "-"}
        </td>

        <td>

            ${getStatusBadge(
                reservation.status
            )}

        </td>

        <td>

            <button
                class="btn btn-view btn-sm"
                onclick="viewReservation(${reservation.id})"
            >

                <i class="bi bi-eye"></i>

                View

            </button>


            <button
                class="btn btn-delete btn-sm ms-2"
                onclick="deleteReservation(${reservation.id})"
            >

                <i class="bi bi-trash"></i>

                Delete

            </button>

        </td>

    `;


    tbody.appendChild(row);

}



// =====================================================
// STATUS BADGE
// =====================================================

function getStatusBadge(
    status
) {

    const value =
        status || "Unknown";


    let badgeClass =
        "bg-secondary";


    if (
        value.toLowerCase() ===
        "reserved"
    ) {

        badgeClass =
            "bg-warning text-dark";

    }


    if (
        value.toLowerCase() ===
        "parked"
    ) {

        badgeClass =
            "bg-primary";

    }


    if (
        value.toLowerCase() ===
        "completed"
    ) {

        badgeClass =
            "bg-success";

    }


    if (
        value.toLowerCase() ===
        "cancelled"
    ) {

        badgeClass =
            "bg-danger";

    }


    return `

        <span class="badge ${badgeClass}">

            ${value}

        </span>

    `;

}



// =====================================================
// VIEW RESERVATION
// =====================================================

async function viewReservation(
    reservationId
) {

    try {

        const token =
            getToken();


        const response =
            await fetch(
                `${API_BASE_URL}/admin_reservations/${reservationId}`,
                {
                    headers: {
                        "Authorization":
                            `Bearer ${token}`
                    }
                }
            );


        if (!response.ok) {

            console.error(
                "View reservation error:",
                response.status
            );

            return;

        }


        const reservation =
            await response.json();


        console.log(
            "Reservation details:",
            reservation
        );


        fillReservationModal(
            reservation
        );


        const modal =
            new bootstrap.Modal(
                document.getElementById(
                    "reservationModal"
                )
            );


        modal.show();


    } catch (error) {

        console.error(
            "View reservation error:",
            error
        );

    }

}



// =====================================================
// FILL MODAL
// =====================================================

// function fillReservationModal(
//     reservation
// ) {

//     document.getElementById(
//         "viewReservationId"
//     ).textContent =
//         reservation.id ?? "-";


//     document.getElementById(
//         "viewStatus"
//     ).innerHTML =
//         getStatusBadge(
//             reservation.status
//         );


//     document.getElementById(
//         "viewUserName"
//     ).textContent =
//         reservation.user_name ?? "-";


//     document.getElementById(
//         "viewUserEmail"
//     ).textContent =
//         reservation.user_email ?? "-";


//     document.getElementById(
//         "viewParkingName"
//     ).textContent =
//         reservation.parking_name ?? "-";


//     document.getElementById(
//         "viewParkingAddress"
//     ).textContent =
//         reservation.parking_address ?? "-";


//     document.getElementById(
//         "viewSlot"
//     ).textContent =
//         reservation.slot_number ?? "-";


//     document.getElementById(
//         "viewVehicleNumber"
//     ).textContent =
//         reservation.vehicle_number ?? "-";


//     document.getElementById(
//         "viewVehicleType"
//     ).textContent =
//         reservation.vehicle_type ?? "-";


//     document.getElementById(
//         "viewReservationDate"
//     ).textContent =
//         formatDate(
//             reservation.reservation_date
//         );


//     document.getElementById(
//         "viewEntryTime"
//     ).textContent =
//         formatDateTime(
//             reservation.entry_time
//         );


//     document.getElementById(
//         "viewExitTime"
//     ).textContent =
//         formatDateTime(
//             reservation.exit_time
//         );


//     document.getElementById(
//         "viewTotalAmount"
//     ).textContent =
//         reservation.total_amount != null
//             ? `₹${reservation.total_amount}`
//             : "-";


//     document.getElementById(
//         "viewCreatedAt"
//     ).textContent =
//         formatDateTime(
//             reservation.created_at
//         );

// }

function fillReservationModal(data) {

    console.log("Reservation details:", data);

    // =========================================
    // RESERVATION
    // =========================================

    document.getElementById("viewReservationId").textContent =
        data.id ?? "-";

    document.getElementById("viewStatus").innerHTML =
        getStatusBadge(data.status);


    // =========================================
    // USER
    // =========================================

    document.getElementById("viewUserName").textContent =
        data.user?.name ??
        data.user?.username ??
        data.user?.email ??
        data.user_name ??
        "-";

    document.getElementById("viewUserEmail").textContent =
        data.user?.email ??
        data.user_email ??
        "-";


    // =========================================
    // PARKING
    // =========================================

    document.getElementById("viewParkingName").textContent =
        data.parking?.parking_name ??
        data.parking?.name ??
        data.parking_name ??
        "-";

    document.getElementById("viewParkingAddress").textContent =
        data.parking?.address ??
        data.parking_address ??
        "-";


    // =========================================
    // SLOT
    // =========================================

    document.getElementById("viewSlot").textContent =
        data.slot?.slot_number ??
        data.slot?.slot_name ??
        data.slot?.name ??
        data.slot?.number ??
        data.slot_number ??
        "-";


    // =========================================
    // VEHICLE
    // =========================================

    document.getElementById("viewVehicleNumber").textContent =
        data.vehicle?.vehicle_number ??
        data.vehicle?.number ??
        data.vehicle_number ??
        "-";

    document.getElementById("viewVehicleType").textContent =
        data.vehicle?.vehicle_type ??
        data.vehicle?.type ??
        data.vehicle_type ??
        "-";


    // =========================================
    // RESERVATION DATE
    // =========================================

    document.getElementById("viewReservationDate").textContent =
    formatDate(
        data.booking_date ??
        data.reservation_date ??
        data.date
    );


    // =========================================
    // ENTRY TIME
    // =========================================

    document.getElementById("viewEntryTime").textContent =
        formatDateTime(
            data.entry_time
        );


    // =========================================
    // EXIT TIME
    // =========================================

    document.getElementById("viewExitTime").textContent =
        formatDateTime(
            data.exit_time
        );


    // =========================================
    // TOTAL AMOUNT
    // =========================================

    document.getElementById("viewTotalAmount").textContent =
        data.total_amount != null
            ? `₹${data.total_amount}`
            : "-";


    // =========================================
    // CREATED AT
    // =========================================

    document.getElementById("viewCreatedAt").textContent =
        formatDateTime(
            data.created_at
        );
}

// =====================================================
// FORMAT DATE
// =====================================================

// function formatDate(
//     value
// ) {

//     if (!value) {
//         return "-";
//     }


//     const date =
//         new Date(value);


//     if (isNaN(date)) {
//         return value;
//     }


//     return date.toLocaleDateString(
//         "en-IN"
//     );

// }

function formatDate(value) {

    if (!value) {
        return "-";
    }

    let dateString = String(value);

    if (
        !dateString.endsWith("Z") &&
        !dateString.includes("+") &&
        !/[+-]\d{2}:\d{2}$/.test(dateString)
    ) {
        dateString = dateString.replace(" ", "T") + "Z";
    }

    const date = new Date(dateString);

    if (isNaN(date.getTime())) {
        return value;
    }

    return date.toLocaleDateString("en-IN", {
        timeZone: "Asia/Kolkata",
        day: "2-digit",
        month: "2-digit",
        year: "numeric"
    });
}



// =====================================================
// FORMAT DATE TIME
// =====================================================

// function formatDateTime(
//     value
// ) {

//     if (!value) {
//         return "-";
//     }


//     const date =
//         new Date(value);


//     if (isNaN(date)) {
//         return value;
//     }


//     return date.toLocaleString(
//         "en-IN"
//     );

// }

function formatDateTime(value) {

    if (!value) {
        return "-";
    }

    let dateString = String(value);

    // If backend sends UTC without timezone information
    if (
        !dateString.endsWith("Z") &&
        !dateString.includes("+") &&
        !/[+-]\d{2}:\d{2}$/.test(dateString)
    ) {
        dateString = dateString.replace(" ", "T") + "Z";
    }

    const date = new Date(dateString);

    if (isNaN(date.getTime())) {
        return value;
    }

    return date.toLocaleString("en-IN", {
        timeZone: "Asia/Kolkata",
        day: "2-digit",
        month: "2-digit",
        year: "numeric",
        hour: "2-digit",
        minute: "2-digit",
        second: "2-digit",
        hour12: true
    });
}


// =====================================================
// SEARCH
// =====================================================

function searchReservations() {

    const input =
        document.getElementById(
            "searchReservation"
        );


    const value =
        input.value
            .trim()
            .toLowerCase();


    if (!value) {

        displayReservations(
            allReservations
        );

        return;

    }


    const filtered =
        allReservations.filter(
            reservation => {

                return (

                    String(
                        reservation.id ?? ""
                    )
                    .toLowerCase()
                    .includes(value)

                    ||

                    String(
                        reservation.user_name ?? ""
                    )
                    .toLowerCase()
                    .includes(value)

                    ||

                    String(
                        reservation.parking_name ?? ""
                    )
                    .toLowerCase()
                    .includes(value)

                    ||

                    String(
                        reservation.slot_number ?? ""
                    )
                    .toLowerCase()
                    .includes(value)

                    ||

                    String(
                        reservation.vehicle_number ?? ""
                    )
                    .toLowerCase()
                    .includes(value)

                );

            }
        );


    displayReservations(
        filtered
    );

}



// =====================================================
// DELETE RESERVATION
// =====================================================

async function deleteReservation(
    reservationId
) {

    const confirmDelete =
        confirm(
            `Delete reservation #${reservationId}?`
        );


    if (!confirmDelete) {
        return;
    }


    const token =
        getToken();


    try {

        const response =
            await fetch(
                `${API_BASE_URL}/admin_reservations/${reservationId}`,
                {
                    method: "DELETE",

                    headers: {
                        "Authorization":
                            `Bearer ${token}`
                    }
                }
            );


        if (!response.ok) {

            console.error(
                "Delete error:",
                await response.text()
            );

            return;

        }


        console.log(
            "Reservation deleted:",
            reservationId
        );


        removeReservation(
            reservationId
        );


    } catch (error) {

        console.error(
            "Delete reservation error:",
            error
        );

    }

}



// =====================================================
// REMOVE RESERVATION FROM UI
// =====================================================

function removeReservation(
    reservationId
) {

    allReservations =
        allReservations.filter(
            reservation =>
                Number(reservation.id) !==
                Number(reservationId)
        );


    displayReservations(
        allReservations
    );

}



// =====================================================
// WEBSOCKET
// =====================================================

function connectWebSocket() {

    socket =
        new WebSocket(
            "ws://smart-parking-system-tz4z.onrender.com/admin_reservations/ws"
        );


    socket.onopen =
        function() {

            console.log(
                "Admin reservation WebSocket connected"
            );


            document.getElementById(
                "socketStatus"
            ).textContent =
                "Live";


            document.getElementById(
                "socketStatus"
            ).className =
                "badge bg-success";


            socket.send(
                "admin_connected"
            );

        };


    socket.onmessage =
        function(event) {

            try {

                const message =
                    JSON.parse(
                        event.data
                    );


                console.log(
                    "Reservation WebSocket:",
                    message
                );


                if (
                    message.event ===
                    "reservation_created"
                ) {

                    handleReservationCreated(
                        message.data
                    );

                }


                else if (
                    message.event ===
                    "reservation_updated"
                ) {

                    handleReservationUpdated(
                        message.data
                    );

                }


                else if (
                    message.event ===
                    "reservation_deleted"
                ) {

                    handleReservationDeleted(
                        message.data
                    );

                }


            } catch (error) {

                console.error(
                    "WebSocket message error:",
                    error
                );

            }

        };


    socket.onerror =
        function(error) {

            console.error(
                "Reservation WebSocket error:",
                error
            );

            document.getElementById(
                "socketStatus"
            ).textContent =
                "Disconnected";


            document.getElementById(
                "socketStatus"
            ).className =
                "badge bg-danger";

        };


    socket.onclose =
        function() {

            console.log(
                "Reservation WebSocket disconnected"
            );


            document.getElementById(
                "socketStatus"
            ).textContent =
                "Reconnecting...";


            document.getElementById(
                "socketStatus"
            ).className =
                "badge bg-warning text-dark";


            setTimeout(
                connectWebSocket,
                3000
            );

        };

}



// =====================================================
// NEW RESERVATION
// =====================================================

function handleReservationCreated(
    reservation
) {

    console.log(
        "New reservation:",
        reservation
    );


    const exists =
        allReservations.some(
            item =>
                Number(item.id) ===
                Number(reservation.id)
        );


    if (exists) {
        return;
    }


    allReservations.unshift(
        reservation
    );


    displayReservations(
        allReservations
    );

}



// =====================================================
// UPDATED RESERVATION
// =====================================================

function handleReservationUpdated(
    reservation
) {

    const index =
        allReservations.findIndex(
            item =>
                Number(item.id) ===
                Number(reservation.id)
        );


    if (index !== -1) {

        allReservations[index] =
            reservation;

    }


    else {

        allReservations.unshift(
            reservation
        );

    }


    displayReservations(
        allReservations
    );

}



// =====================================================
// DELETED RESERVATION
// =====================================================

function handleReservationDeleted(
    data
) {

    const reservationId =
        data.id ??
        data.reservation_id;


    removeReservation(
        reservationId
    );

}



// =====================================================
// ENTER KEY SEARCH
// =====================================================

document
    .getElementById(
        "searchReservation"
    )
    ?.addEventListener(
        "keyup",
        function(event) {

            if (
                event.key === "Enter"
            ) {

                searchReservations();

            }

        }
    );



// =====================================================
// PAGE LOAD
// =====================================================

document.addEventListener(
    "DOMContentLoaded",
    function() {

        loadReservations();

        connectWebSocket();

    }
);

// --------------delete----------------------\
async function deleteReservation(reservationId) {

    const confirmDelete = confirm(
        `Are you sure you want to delete Reservation #${reservationId}?`
    );

    if (!confirmDelete) {
        return;
    }

    try {

        const response = await fetch(
            `${API_URL}/owner/reservations/${reservationId}`,
            {
                method: "DELETE",

                headers: {
                    "Authorization": `Bearer ${token}`
                }
            }
        );

        // -----------------------------------------
        // Safely read response
        // -----------------------------------------

        const text = await response.text();

        let data = {};

        try {
            data = text ? JSON.parse(text) : {};
        }
        catch (jsonError) {

            console.error(
                "Invalid JSON response:",
                text
            );
        }

        console.log(
            "Delete status:",
            response.status
        );

        console.log(
            "Delete response:",
            data
        );

        // -----------------------------------------
        // ERROR
        // -----------------------------------------

        if (!response.ok) {

            alert(
                data.detail ||
                `Delete failed (${response.status})`
            );

            return;
        }

        // -----------------------------------------
        // SUCCESS
        // -----------------------------------------

        alert(
            "Reservation deleted successfully"
        );

        // -----------------------------------------
        // Remove row immediately
        // -----------------------------------------

        const row = document.querySelector(
            `tr[data-id="${reservationId}"]`
        );

        if (row) {
            row.remove();
        }

    }
    catch (error) {

        console.error(
            "Delete error:",
            error
        );

        alert(
            "Unable to delete reservation."
        );
    }
}